### DISCUSSION TITLE
TestStore

### DISCUSSION THREAD

**Discussion by WHuangz88** [pointfreeco/swift-composable-architecture#1213] [Q&A]
<img width="507" alt="Screen Shot 2022-07-16 at 01 22 36" src="https://user-images.githubusercontent.com/95329209/179287680-a47bf2f9-5604-4104-91db-74b5cb20bee0.png">

<img width="702" alt="Screen Shot 2022-07-16 at 01 26 28" src="https://user-images.githubusercontent.com/95329209/179287692-6ef0bb88-d9ef-49aa-98bf-068cc0fd2be4.png">

Hi guys, does anyone know how to solve this error ? _Expected to received an action, but received none_. I found out that the problem is because when I trigger the `.send(.loadNews)` method nothing was changed in my reducer state, since it's just used for api call. does any know how to fix it?

**Comment by mbrandonw** [comment]
Can you share more of what the `getNews` implementation looks like in your code? Are you using a mocked version of the endpoint in tests?

**Reply by WHuangz88** [reply]
```
func getNews(req: PagingRequest) -> Single<StockNews> {
        return self.repo.getNews(page: req.page,
                                 pageSize: req.pageSize,
                                 query: "")
    }
```

so `getNews` is just a normal API call, and in test file I mocked it

**Reply by mbrandonw** [reply]
Somehow this effect is not emitting a value. Can you put a `.print()` on the `self.repo.getNews` in order to see if something is actually emitting?

I also see that you are using a `Single`, which must mean you aren't using this library but instead either the Rx or ReactiveCocoa fork of the library? If this is true then perhaps there is a bug in one of those libraries.

**Reply by WHuangz88** [reply]
the data I received when calling self.repo.getNews is not from actual API call but from dummy data that I already provide from `Stocknews.mock`. When I tried to debug my reducer when running the code, it actually triggers the `.loadNews` action and return the same data as what being mocked

**Reply by WHuangz88** [reply]
I found the root cause of the problem because I put `.observeOn(MainScheduler.asyncInstance)` in the store effect when calling `func send(_ action: Action)` to prevent

- * The store was sent an action recursively. This can occur when you run an effect \
          directly in the reducer, rather than returning it from the reducer. Check the stack (⌘7) \
          to find frames corresponding to one of your reducers. That code should be refactored to \
          not invoke the effect directly.
-* The store has been sent actions from multiple threads. The `send` method is not \
          thread-safe, and should only ever be used from a single thread (typically the main \
          thread). Instead of calling `send` from multiple threads you should use effects to \
          process expensive computations on background threads so that it can be fed back into the \
          store.
          
do you have any ideas on how to prevent it without having to put this `.observeOn(MainScheduler.asyncInstance)`

**Reply by mbrandonw** [reply]
I'm not sure I can help because 1) it seems you are using a different library since `.observeOn` is not an operator in Combine, and 2) the message you are seeing "The store was sent an action recursively..." is from a very, very old version of the library and is no longer around.

Which fork of TCA are you using?

**Reply by WHuangz88** [reply]
from [here](https://github.com/dannyhertz/rxswift-composable-architecture) , since i m using rx swift 😄

**Reply by mbrandonw** [reply]
That library is very out of date with this library so it's hard to help unfortunately. The assertion failure complaining about recursive actions was removed a long time ago.
